lib/mapping/CCampaignHandler.cpp

```diff
@@ -151,45 +201,292 @@ std::string CCampaignHandler::readLocalizedString(CBinaryReader & reader, std::s
+		ret.hasPrologEpilog = !identifier.isNull();
+		if(ret.hasPrologEpilog)
+		{
+			ret.prologVideo = identifier["video"].String();
+			ret.prologMusic = identifier["music"].String();
+			ret.prologText = identifier["text"].String();
+		}
+		return ret;
+	};
+
+	CCampaignScenario ret;
+	ret.conquered = false;
+	ret.mapName = reader["map"].String();
+	for(auto & g : reader["preconditions"].Vector())
+		ret.preconditionRegions.insert(g.Integer());
+
+	ret.regionColor = reader["color"].Integer();
+	ret.difficulty = reader["difficulty"].Integer();
+	ret.regionText = reader["regionText"].String();
+	ret.prolog = prologEpilogReader(reader["prolog"]);
+	ret.epilog = prologEpilogReader(reader["epilog"]);
+
+	ret.travelOptions = readScenarioTravelFromJson(reader);
+
+	return ret;
+}
+
+CScenarioTravel CCampaignHandler::readScenarioTravelFromJson(JsonNode & reader)
+{
+	CScenarioTravel ret;
+
+	std::map<std::string, ui8> startOptionsMap = {
+		{"none", 0},
+		{"bonus", 1},
+		{"crossover", 2},
+		{"hero", 3}
+	};
+	
+	std::map<std::string, CScenarioTravel::STravelBonus::EBonusType> bonusTypeMap = {
+		{"spell", CScenarioTravel::STravelBonus::EBonusType::SPELL},
+		{"creature", CScenarioTravel::STravelBonus::EBonusType::MONSTER},
+		{"building", CScenarioTravel::STravelBonus::EBonusType::BUILDING},
+		{"artifact", CScenarioTravel::STravelBonus::EBonusType::ARTIFACT},
+		{"scroll", CScenarioTravel::STravelBonus::EBonusType::SPELL_SCROLL},
+		{"primarySkill", CScenarioTravel::STravelBonus::EBonusType::PRIMARY_SKILL},
+		{"secondarySkill", CScenarioTravel::STravelBonus::EBonusType::SECONDARY_SKILL},
+		{"resource", CScenarioTravel::STravelBonus::EBonusType::RESOURCE},
+		//{"prevHero", CScenarioTravel::STravelBonus::EBonusType::HEROES_FROM_PREVIOUS_SCENARIO},
+		//{"hero", CScenarioTravel::STravelBonus::EBonusType::HERO},
+	};
+	
+	std::map<std::string, ui32> primarySkillsMap = {
+		{"attack", 0},
+		{"defence", 8},
+		{"spellpower", 16},
+		{"knowledge", 24},
+	};
+	
+	std::map<std::string, ui16> heroSpecialMap = {
+		{"strongest", 0xFFFD},
+		{"generated", 0xFFFE},
+		{"random", 0xFFFF}
+	};
+	
+	std::map<std::string, ui8> resourceTypeMap = {
+		//FD - wood+ore
+		//FE - mercury+sulfur+crystal+gem
+		{"wood", 0},
+		{"mercury", 1},
+		{"ore", 2},
+		{"sulfur", 3},
+		{"crystal", 4},
+		{"gems", 5},
+		{"gold", 6},
+		{"common", 0xFD},
+		{"rare", 0xFE}
+	};
+	
+	for(auto & k : reader["heroKeeps"].Vector())
+	{
+		if(k.String() == "experience") ret.whatHeroKeeps.experience = true;
+		if(k.String() == "primarySkills") ret.whatHeroKeeps.primarySkills = true;
+		if(k.String() == "secondarySkills") ret.whatHeroKeeps.secondarySkills = true;
+		if(k.String() == "spells") ret.whatHeroKeeps.spells = true;
+		if(k.String() == "artifacts") ret.whatHeroKeeps.artifacts = true;
+	}
+	
+	for(auto & k : reader["keepCreatures"].Vector())
+	{
+		if(auto identifier = VLC->modh->identifiers.getIdentifier(CModHandler::scopeMap(), "creature", k.String()))

```

Comment:
Keep in mind that map scope is not implemented and aliased to "game" scope.
Meaning, it will resolve identifiers even if they are not from dependent mod.

Not implemented since maps didn't have list of dependent mods in its format.

Reply:
I added it into vmap in this PR. Could you please check changes in CMap and CMapService? Maybe I did it not in supposed way

Reply:
ModHandler still needs to use information provided by CMap / CMapService. This would need following changes:
- CModHandler::scopeMap should return "map" instead of "game"
- CIdentifierStorage::getPossibleIdentifiers should (in one way or another) request.localScope "map" and add to allowedScopes all map dependencies
- Old map versions should somehow still use "game" scope?

Up to you whether to do these changes now or not.

Reply:
I already have too many changes in this branch, I'll merge to proceed with other improvements